src/NULS/Signer.cpp

```diff
@@ -0,0 +1,149 @@
+    data.push_back(static_cast<uint8_t>((val >> 40)));
+}
+
+static inline void serializerRemark(std::string& remark, std::vector<uint8_t>& data)
+{
+    writeCompactSize(remark.length(), data);
+    std::copy(remark.begin(), remark.end(), std::back_inserter(data));
+}
+
+static inline void serializerInput(std::vector<Proto::TransactionInput>& inputs, std::vector<uint8_t>& data)
+{
+    writeCompactSize(inputs.size(), data);
+    for (const auto& input : inputs) {
+        Data owner = parse_hex(input.from_hash());
+        // Input owner is txid and index
+        writeCompactSize((uint16_t) input.from_index(), owner);
+
+        writeCompactSize(owner.size(), data);
+        std::copy(owner.begin(), owner.end(), std::back_inserter(data));
+
+        encode64LE((uint64_t) input.amount(), data);
+        encode48LE((uint64_t) input.lock_time(), data);
+    }
+}
+
+static inline void serializerOutput(std::vector<Proto::TransactionOutput>& outputs, std::vector<uint8_t>& data)
+{
+    writeCompactSize(outputs.size(), data);
+    for (const auto& output : outputs) {
+        const auto& toAddress = output.to_address();
+        if (!NULS::Address::isValid(toAddress)) {
+            throw std::invalid_argument("Invalid address");
+        }
+
+        const auto& addr = NULS::Address(toAddress);
+        writeCompactSize(addr.bytes.size()-1, data);
+        std::copy(addr.bytes.begin(), addr.bytes.end()-1, std::back_inserter(data));
+        encode64LE((uint64_t) output.amount(), data);
+        encode48LE((uint64_t) output.lock_time(), data);
+    }
+}
+
+static inline Data calcTransactionDigest(Data& data)
+{
+    Data hash1 = Hash::sha256(data);
+    Data hash2 = Hash::sha256(hash1);
+    return hash2;
+}
+
+static inline Data makeTransactionSignature(PrivateKey& privateKey, Data& txHash)
+{
+    PublicKey pubKey = privateKey.getPublicKey(PublicKeyType::secp256k1);
+
+    Data transactionSignature = Data();
+    writeCompactSize(pubKey.bytes.size(), transactionSignature);
+    std::copy(pubKey.bytes.begin(), pubKey.bytes.end(), std::back_inserter(transactionSignature));
+    auto signature = privateKey.signAsDER(txHash, TWCurve::TWCurveSECP256k1);
+
+    transactionSignature.push_back(static_cast<uint8_t>(0x00));
+    writeCompactSize(signature.size(), transactionSignature);
+
+    std::copy(signature.begin(), signature.end(), std::back_inserter(transactionSignature));
+    return transactionSignature;
+}
+
+std::vector<uint8_t> Signer::sign(uint64_t timestamp) const
+{
+    if (tx.private_key().empty()) {
+        throw std::invalid_argument("Must have private key string");
+    }
+    else if (tx.inputs_size()==0) {
+        throw std::invalid_argument("Not enough input balance to do the transaction");
+    }
+    else if (tx.outputs_size()==0) {
+        throw std::invalid_argument("There must be at least one output, something is missed");
+    }
+    auto priv = Address::importHexPrivateKey(tx.private_key());
```

Comment:
I don't think we need `importHexPrivateKey`, you can add a `private_key` to `SigningInput` and get it from signer

Reply:
There are 3 possibilities for the private key exported by nuls wallet. If the import is used directly, it will fail because the length does not meet the requirements.

Reply:
let's just consider the simplest case that we derive a private key from mnemonic, @vikmeup what do you think?